src/utils.jl

```diff
@@ -50,50 +50,30 @@ accuracy(nlp::T) where {T <: AbstractKnetNLPModel} =
   Knet.accuracy(nlp.chain; data = nlp.minibatch_test)
 
 """
-    build_layer_from_vec(v :: Vector{T}, var_layers :: CuArray{T, N, CUDA.Mem.DeviceBuffer} where N, index :: Int) where {T <: Number}
-
-Inverse of the function `Knet.cat1d`; build a `CuArray` similar to `var_layers` from `v`.
-The return values are those of the vector `v` in the range `index+1:index+consumed_index`.
-This method is not optimized; it allocates memory.
-"""
-function build_layer_from_vec(
-  v::Vector{T},
-  var_layers::CuArray{T, N, CUDA.Mem.DeviceBuffer} where {N},
-  index::Int,
-) where {T <: Number}
-  dims = ndims(var_layers)
-  size_var_layers = size(var_layers)
-  tmp_array = Array{T, dims}(undef, size_var_layers)
-  cuArray = CuArray(tmp_array)
-  product_dims = build_layer_from_vec!(cuArray, v, index)
-  return (cuArray, product_dims)
-end
+    build_layer_from_vec!(array :: A, v :: V, index :: Int) where {T <: Number, N, A<:AbstractArray{T,N}, V<:AbstractVector{T}}
 
-"""
-    build_layer_from_vec!(cuArray :: CuArray{T, N, CUDA.Mem.DeviceBuffer} where {N}, v :: Vector{T}, index :: Int) where {T <: Number}
-
-Inverse of the function `Knet.cat1d`; set `cuArray` to the values of `v` in the range `index+1:index+consumed_index`.
+Inverse of the function `Knet.cat1d`; set `array` to the values of `v` in the range `index+1:index+consumed_index`.
 """
 function build_layer_from_vec!(
-  cuArray::CuArray{T, N, CUDA.Mem.DeviceBuffer} where {N},
-  v::Vector{T},
+  array::A,
+  v::V,
   index::Int,
-) where {T <: Number}
-  sizecuArray = reduce(*, size(cuArray))
-  copyto!(cuArray, v[(index + 1):(index + sizecuArray)])
-  return sizecuArray
+) where {T <: Number, V<:AbstractVector{T}, N, A<:AbstractArray{T, N}}
```

Comment:
This still looks complicated. Couldn't it be `array::AbstractArray{T, N}, v::AbstractVector{T} ... where {T <: Number, N}`?

Reply:
fixed